### PR TITLE
Add Gemma Rotary Embedding

### PR DESCRIPTION
**Pull request by jingyanwangms**
### Description
Add GemmaRotaryEmbedding kernel which includes sin and cos in GemmaRotaryEmbedding forward and apply_rotary_pos_emb. See gemma_rotary_emb_impl.cu for subgraph details

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/gemma_rotary_emb_impl.cu [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    const T* q_rot, 
+    const T* k, 
+    const T* k_rot, 
+    const int batch_size,
+    const int num_heads,
+    const int seq_len,
+    const int dim
+    ) {
+
+  const dim3 block(dim);
```
Comment (wschin) [inline onnxruntime/contrib_ops/cuda/bert/gemma_rotary_emb_impl.cu:80]: How do you decide the value? Is it copied from other element-wise kernels?

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 2]
```diff
@@ -1214,6 +1214,50 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
           propagateShapeFromInputToOutput(ctx, 0, 0);
         }));
 
+constexpr const char* GemmaRotaryEmbedding_ver1_doc = R"DOC(
+GemmaRotaryEmbedding is the implementation of rotary positional embeddings (RoPE). It includes sin and cos in GemmaRotaryEmbedding forward and apply_rotary_pos_emb
```
Comment (wschin) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1218]: Write done the math formulation for this kernel. E.g.,

def gemma_rotary_embedding(x, y, z):
  w = Sin(x)
  z = Cast(y, to=?) # for fp16 mix-precision training. Other types are not supported.
  ...

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 3]
```diff
@@ -1214,6 +1214,50 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
...
+)DOC";
+ONNX_MS_OPERATOR_SET_SCHEMA(
+    GemmaRotaryEmbedding, 1,
+    OpSchema()
+        .SetDoc(GemmaRotaryEmbedding_ver1_doc)
+        .Input(0,
+               "emb",
+               "embeddding - 3D tensor with shape (batch_size, seq_len, dim)",
```
Comment (wschin) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1226]: Great to see shapes mentioned in document. Thanks.

**Comment by tianleiwu** [comment]
There is no need to add such operator since RoPE is supported in GroupQueryAttention operator. That will have better performance.

We have optimization for Gemma. Try it out: https://onnxruntime.ai/docs/genai/, and take a look at https://onnxruntime.ai/docs/genai/howto/build-model.html about converting Gemma to onnx model.

**Comment by wschin** [comment]
> There is no need to add such operator since RoPE is supported in GroupQueryAttention operator. That will have better performance since there is less intermediate data I/O and less kernel launches.
> 
> We have optimization for Gemma. Try it out: https://onnxruntime.ai/docs/genai/, and take a look at https://onnxruntime.ai/docs/genai/howto/build-model.html about converting Gemma to onnx model.

Two major blockers of using those contrib ops is --- 1) no or very insufficient documentation (need to detailed enough such that backward formulation can be derived), 2) they are not designed for training (e.g., no dropout support in attention, the area training wants to fuse doesn't always match the area inference wants to fuse, not enough document to even figure out what is missing, etc). Until the two blockers are resolved, I'd treat training op independently than inference op. Does it make sense?

File: onnxruntime/contrib_ops/cuda/bert/gemma_rotary_emb_impl.cu (hunk 1 above)
Comment (jingyanwangms) [reply onnxruntime/contrib_ops/cuda/bert/gemma_rotary_emb_impl.cu:80]: No. I only thought it's easier to have grid dim3 corresponds to [batch_size, num_heads, seq_len, dim] because I calculated index instead of doing unsqueeze. Other element wise kernel (I'm looking at bias_gelu) seems launch kernel with kElementsPerThread. Is that what you're thinking? Now to think about it, I can probably do the same.

Comment (jingyanwangms) [reply onnxruntime/contrib_ops/cuda/bert/gemma_rotary_emb_impl.cu:80]: Working on this. Resolved other issues first

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 4]
```diff
@@ -1215,7 +1215,23 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
         }));
 
 constexpr const char* GemmaRotaryEmbedding_ver1_doc = R"DOC(
-GemmaRotaryEmbedding is the implementation of rotary positional embeddings (RoPE). It includes sin and cos in GemmaRotaryEmbedding forward and apply_rotary_pos_emb
+GemmaRotaryEmbedding is the implementation of below part of rotary positional embeddings (RoPE). It implements below from modeling_gemma.py.
+- Only support fp16, other types are not support
+- rotate_half is not included
+
+class LlamaRotaryEmbedding(nn.Module):
```
Comment (wschin) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1222]: This doesn't explain the math. For example, what are inputs/outputs of this op? Are they inputs/outputs of `LlamaRotaryEmbedding`? Or are they inputs/outputs of `apply_rotary_pos_emb`? Do you write two fused kernels, one for `LlamaRotaryEmbedding` and the other one for `apply_rotary_pos_emb`? Let's consider fusion of MatMul and Add as an example. I'd using onnxscript syntax to define the fused sub-graph in Python:
```
# The computation graph below (defined by onnxscript) is the math representation of this kernel.
# This kernel has 3 inputs, `x`, `y`, and `bias`. This kernel produces only one output tensor.
onnxscript.script()
def matmul_add_fusion(x, y, bias):
  z = op.MatMul(x, y)
  w = op.Add(z, bias)
  return w
```

**Comment by kunal-vaishnavi** [comment]
> There is no need to add such operator since RoPE is supported in GroupQueryAttention operator. That will have better performance since there is less intermediate data I/O and less kernel launches.
> 
> We have optimization for Gemma. Try it out: https://onnxruntime.ai/docs/genai/, and take a look at https://onnxruntime.ai/docs/genai/howto/build-model.html about converting Gemma to onnx model.

There is also already a standalone `RotaryEmbedding` op that works for many model architectures including Gemma.

https://github.com/microsoft/onnxruntime/blob/e6ca360695699e73fc5acd8830f6bb14609b7ef6/onnxruntime/core/graph/contrib_ops/bert_defs.cc#L1166-L1215

Implementations:
- [CPU](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/contrib_ops/cpu/bert/rotary_embedding.cc)
- [CUDA](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/contrib_ops/cuda/bert/rotary_embedding.cc)
- [DML](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorRotaryEmbedding.cpp)
- [Web](https://github.com/microsoft/onnxruntime/blob/main/js/web/lib/wasm/jsep/webgpu/ops/rotary-embedding.ts)

Test cases that show fusions and op usage:

- [Parity tests in C++](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/contrib_ops/rotary_embedding_op_test.cc)
- [Parity tests in Python](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/python/transformers/test_parity_rotary_embedding.py)
- [RotaryEmbedding + MultiHeadAttention fusion](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/python/transformers/test_rotary_mha_fusion.py)
- [RotaryEmbedding fusion](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/test/python/transformers/test_rotary_embedding_fusion.py)

If there are training-specific changes needed, I think making them in the existing `RotaryEmbedding` op would be better.

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc (hunk 4 above)
Comment (jingyanwangms) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc]: I see what you mean. Fied

File: onnxruntime/contrib_ops/cuda/bert/gemma_rotary_emb_impl.cu (hunk 1 above)
Comment (jingyanwangms) [reply onnxruntime/contrib_ops/cuda/bert/gemma_rotary_emb_impl.cu]: Fixed

File: onnxruntime/core/graph/contrib_ops/bert_defs.cc [hunk 5]
```diff
@@ -1214,6 +1214,62 @@ ONNX_MS_OPERATOR_SET_SCHEMA(
...
 
+constexpr const char* GemmaRotaryEmbedding_ver1_doc = R"DOC(
+GemmaRotaryEmbedding is the implementation of below part of rotary positional embeddings (RoPE). It implements below from modeling_gemma.py.
+
+def gemma_rotary_embedding(emb, q, q_rot, k, k_rot):
+  sin_val = Sin(emb)
+  casted_sin = Cast(sin_val, to=fp16) # for fp16 mix-precision training. Other types are not supported.
+  cos_val = Cos(emb)
+  casted_cos = Cast(cos_val, to=fp16)
+  unsqueezed_sin = Unsqueeze(casted_sin)
```
Comment (wschin) [inline onnxruntime/core/graph/contrib_ops/bert_defs.cc:1225]: I assume Unsqueeze has two inputs. Where is the 2nd one?

Comment (jingyanwangms) [reply onnxruntime/core/graph/contrib_ops/bert_defs.cc:1225]: Current implementation checks and assumes the dimensions of input so unsqueeze_axis is not necessary. Should I add unsqueeze_axis as input?

**Review by wschin — APPROVED** [review]
Please move this op to training folder afterwards. Thanks.

**Review by wschin — DISMISSED** [review]
Thanks.

**Review by wschin — APPROVED** [review]
